### PR TITLE
Knuth-Morris algo readme

### PR DESCRIPTION
**Pull request by raksha009**
Fixes #2555 

#### Checklist:
- [ ] 4 space indentation.
- [ ] Coding conventions are followed.
- [ ] Input is taken dynamically.
- [ ] Sample Input / Output is added at the end of file.
- [X] Logic Documentation (Comments).
- [X] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- ADDED NEW FILE
- 

#### Languages Used:
- 
- 

#### Files Added:
- Knuth-Morris_Pratt_Algorithm_README.md
- 

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by raksha009** [comment]
plz review my pr @asha15  @singh-shreya6 maam

**Comment by raksha009** [comment]
Plz review my Pr @singh-shreya6 @asha15 @raghav-dalmia @MastersAbh

File: Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
+## Knuth-Morris Pratt Algorithm
+
+1. It was invented by Donald Knuth and Vaughan Pratt.
+2. It was the first linear time complexity algorithm for string matching which is used to find a pattern in a text.
+3. This algorithm compares character by character from left to right.
+
+
```
Comment (asha15) [inline Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md:7]: please remove this extra lines

File: Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+**Components of Knuth-Morris Pratt algorithm**
```
Comment (asha15) [inline Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md:9]: don't bold it. use h2 header ( ## )

File: Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
+## Knuth-Morris Pratt Algorithm
```
Comment (asha15) [inline Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md:1]: keep one #

File: Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+
+
+**Components of Knuth-Morris Pratt algorithm**
+
+Prefix Table : In this algorithm whenever a mismatch occurs,it uses a preprocessed table known as Prefix Table to skip characters comparison while matching.
+
+LPS Table : It stands for Longest proper Prefix which is also suffix. Sometimes prefix table is also known as LPS Table.
+
+
```
Comment (asha15) [inline Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md:15]: remove the extra lines

File: Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+##### **Steps for Creating LPS Table**
```
Comment (asha15) [inline Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md:17]: remove the bold sign 
keep only two #

File: Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+##### **Steps for Creating LPS Table** 
+
+1. Define a one dimensional array with the size equal to the length of the Pattern. (LPS[size])
+2. Define variables i & j. Set i = 0, j = 1 and LPS[0] = 0.
+3. Compare the characters at Pattern[i] and Pattern[j].
+4. If both are matched then set LPS[j] = i+1 and increment both i & j values by one. Go to Step 3.
+5. If both are not matched then check the value of variable 'i'. If it is '0' then set LPS[j] = 0 and increment 'j' value by one, if it is not '0' then set i = LPS[i-1].  Step 3.
+6. Repeat above steps until all the values of LPS[] are filled. 
+
+**ALGORITHM**
```
Comment (asha15) [inline Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md:26]: make this as a h2 header

File: Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+         print the location (i-j) as there is the pattern
+         j := prefArray[j-1]
+      else if i < n AND pattern[j] ≠ text[i] then
+         if j ≠ 0 then
+            j := prefArray[j - 1]
+         else
+            increase i by 1
+   done
+   
+   End
```
Comment (asha15) [inline Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md:50]: remove the space at the beginning

File: Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+```
+
+**Output**: The location where patterns are found.
+
+ **How The Knuth-Morris Pratt Algorithm Work?**
```
Comment (asha15) [inline Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md:55]: make this as a h2 header

**Comment by raksha009** [comment]
done the changes @asha15 maam

File: Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+## Steps for Creating LPS Table
+
+1. Define a one dimensional array with the size equal to the length of the Pattern. (LPS[size])
+2. Define variables i & j. Set i = 0, j = 1 and LPS[0] = 0.
+3. Compare the characters at Pattern[i] and Pattern[j].
+4. If both are matched then set LPS[j] = i+1 and increment both i & j values by one. Go to Step 3.
+5. If both are not matched then check the value of variable 'i'. If it is '0' then set LPS[j] = 0 and increment 'j' value by one, if it is not '0' then set i = LPS[i-1].  Step 3.
+6. Repeat above steps until all the values of LPS[] are filled. 
+
+ALGORITHM
```
Comment (asha15) [inline Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md:22]: please make this as a sub header

Comment (asha15) [inline Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md:22]: please make this as a sub header

File: Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+Prefix Table : In this algorithm whenever a mismatch occurs,it uses a preprocessed table known as Prefix Table to skip characters comparison while matching.
+
+LPS Table : It stands for Longest proper Prefix which is also suffix. Sometimes prefix table is also known as LPS Table.
+
+## Steps for Creating LPS Table
+
+1. Define a one dimensional array with the size equal to the length of the Pattern. (LPS[size])
+2. Define variables i & j. Set i = 0, j = 1 and LPS[0] = 0.
+3. Compare the characters at Pattern[i] and Pattern[j].
+4. If both are matched then set LPS[j] = i+1 and increment both i & j values by one. Go to Step 3.
```
Comment (singh-shreya6) [inline Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md:18]: Please follow the same coding convention here also. Add space around +

File: Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+**Output**: The location where patterns are found.
+
+##  How The Knuth-Morris Pratt Algorithm Work?
+
+Let us see an example
+
+Consider the following Text and Pattern
+
+​						Text : ABC ABCDAB ABCDABCDABDE
```
Comment (singh-shreya6) [inline Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md:56]: This looks different in actual view. It doesn't come center aligned.

File: Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+​											0123456
+
+Here mismatch occurred at Pattern[6], so we need to consider LPS[5] value. Since LPS[5] value is '2' we compare Pattern[2] character with mismatched character in Text.
+
+**Step 3:** Since LPS value is '2' no need to compare the Pattern[0] and Pattern[1] values
+
+​						Text: ABC ABCDAB ABCDABCDABDE
+
+ 						Pattern:        ABCDABD
```
Comment (singh-shreya6) [inline Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md:88]: This line also looks different from others. Please have a look at actual view of read me

File: Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+**TIME COMPLEXITY**
+
+The time complexity of Knuth-Morris Pratt Algorithm is O(n).
+
+**SPACE COMPLEXITY**
+
+It also has a space complexity of O (m) because there’s some pre-processing involved.
+
+
```
Comment (singh-shreya6) [inline Knuth_Morris_Pratt_Algorithm/Knuth-Morris_Pratt_Algorithm_README.md:124]: Please add References here.

**Comment by raksha009** [comment]
Pl review @asha15 @singh-shreya6 maam. I have done the changes.

**Pull request merged**